pkg/controllermanager/controller/plant/plant_control.go

```diff
@@ -0,0 +1,357 @@
+	updatePlant.Status.Conditions = conditions
+
+	if !equality.Semantic.DeepEqual(plant, updatePlant) {
+		if _, err := c.k8sGardenClient.GardenCore().CoreV1alpha1().Plants(updatePlant.Namespace).UpdateStatus(updatePlant); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *defaultPlantControl) delete(ctx context.Context, plant *gardencorev1alpha1.Plant, logger logrus.FieldLogger) error {
+	if err := c.removeFinalizersFromPlant(plant); err != nil {
+		return err
+	}
+	return c.removeFinalizerFromSecret(ctx, plant)
+}
+
+func (c *defaultPlantControl) removeFinalizerFromSecret(ctx context.Context, plant *gardencorev1alpha1.Plant) error {
+	secret, err := c.secretsLister.Secrets(plant.Namespace).Get(plant.Spec.SecretRef.Name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// secret is already deleted - fine
+			return nil
+		}
+		return err
+	}
+	secretFinalizers := sets.NewString(secret.Finalizers...)
+	secretFinalizers.Delete(FinalizerName)
+	secret.Finalizers = secretFinalizers.UnsortedList()
+	if err := c.k8sGardenClient.Client().Update(ctx, secret); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *defaultPlantControl) removeFinalizersFromPlant(plant *gardencorev1alpha1.Plant) error {
+	updatePlant := plant.DeepCopy()
+	finalizers := sets.NewString(plant.Finalizers...)
+	finalizers.Delete(FinalizerName)
+	updatePlant.Finalizers = finalizers.UnsortedList()
+	if sets.NewString(plant.Finalizers...).Has(FinalizerName) {
+		if _, err := c.k8sGardenClient.GardenCore().CoreV1alpha1().Plants(updatePlant.Namespace).UpdateStatus(updatePlant); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *defaultPlantControl) initializeHealthChecker(plantClusterClient client.Client, discoveryClient *kubernetesclientset.Clientset) *HealthChecker {
+	return NewHealthChecker(plantClusterClient, discoveryClient)
+}
+
+func (c *defaultPlantControl) initializePlantClients(plant *gardencorev1alpha1.Plant, key string, kubeconfigSecretValue []byte) (client.Client, *kubernetesclientset.Clientset, error) {
+	config, err := clientcmd.RESTConfigFromKubeConfig(kubeconfigSecretValue)
+	if err != nil {
+		return nil, nil, fmt.Errorf("%v:%v", "invalid kubconfig supplied resulted in: ", err)
+	}
+	plantClusterClient, err := kubernetes.NewRuntimeClientForConfig(config, client.Options{
+		Scheme: kubernetes.PlantScheme,
+	})
+
+	discoveryClient, err := kubernetesclientset.NewForConfig(config)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return plantClusterClient, discoveryClient, nil
+}
+
+func (c *defaultPlantControl) healthChecks(ctx context.Context, healthChecker *HealthChecker, logger logrus.FieldLogger, apiserverAvailability, healthyNodes gardencorev1alpha1.Condition) (gardencorev1alpha1.Condition, gardencorev1alpha1.Condition) {
+	var wg sync.WaitGroup
+
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		apiserverAvailability = healthChecker.CheckAPIServerAvailability(apiserverAvailability)
```

Comment:
Below you do `healthyNodes = newConditionOrError(healthyNodes, *newNodes, err)` - why is it not needed here?